etl/src/v2/workers/apply.rs

```diff
@@ -173,125 +241,250 @@ impl<S, D> Hook<S, D> {
+        // TODO: switch to a connection pool which hands out connections.
+        let replication_client = self.replication_client.duplicate().await?;
+        let worker = TableSyncWorker::new(
+            self.identity.clone(),
+            self.config.clone(),
+            replication_client,
+            self.pool.clone(),
+            table_id,
+            self.schema_cache.clone(),
+            self.state_store.clone(),
+            self.destination.clone(),
+            self.shutdown_rx.clone(),
+        );
+
+        let mut pool = self.pool.write().await;
+        if let Err(err) = pool.start_worker(worker).await {
+            // TODO: check if we want to build a backoff mechanism for retrying the
+            //  spawning of new table sync workers.
+            error!("Failed to start table sync worker: {}", err);
+
+            return Err(err.into());
+        }
+
+        Ok(())
+    }
+
+    async fn handle_syncing_table(
+        &self,
+        table_id: Oid,
+        current_lsn: PgLsn,
+    ) -> Result<bool, ApplyWorkerHookError> {
+        let table_sync_worker_state = {
+            let pool = self.pool.read().await;
+            pool.get_active_worker_state(table_id)
+        };
+
+        let Some(table_sync_worker_state) = table_sync_worker_state else {
+            info!("Creating new sync worker for table {}", table_id);
+            self.start_table_sync_worker(table_id).await?;
+
+            return Ok(true);
+        };
+
+        self.handle_existing_worker(table_id, table_sync_worker_state, current_lsn)
+            .await
+    }
+
+    async fn handle_existing_worker(
+        &self,
+        table_id: Oid,
+        table_sync_worker_state: TableSyncWorkerState,
+        current_lsn: PgLsn,
+    ) -> Result<bool, ApplyWorkerHookError> {
+        let mut catchup_started = false;
+        {
+            let mut inner = table_sync_worker_state.get_inner().write().await;
+            if inner.replication_phase().as_type() == TableReplicationPhaseType::SyncWait {
+                inner
+                    .set_phase_with(
+                        TableReplicationPhase::Catchup { lsn: current_lsn },
+                        self.state_store.clone(),
+                    )
+                    .await?;
+
+                catchup_started = true;
+            }
+        }
+
+        if catchup_started {
+            let result = table_sync_worker_state
+                .wait_for_phase_type(
+                    TableReplicationPhaseType::SyncDone,
+                    self.shutdown_rx.clone(),
+                )
+                .await;
+
+            // If we are told to shut down while waiting for a phase change, we will signal this to
+            // the caller.
+            if result.should_shutdown() {
+                return Ok(false);
+            }
+
+            info!("Sync completed for table {}", table_id);
+        }
+
+        Ok(true)
+    }
+
+    async fn active_table_replication_states(
```

Comment:
We get all the active replication states here, so that they can be used both during initialization and table syncing.

I am not super happy about this implementation since it would be cool to statically enforce that now we have only active states, but for now it's fine.